File: hot100/Solution0230.cpp
#include <stack>
#include <unordered_map>
using namespace std;
// 给定一个二叉搜索树的根节点 root ，和一个整数 k ，请你设计一个算法查找其中第 k 小的元素（从 1 开始计数）。

// 示例 1：

// 输入：root = [3,1,4,null,2], k = 1
// 输出：1
// 示例 2：

// 输入：root = [5,3,6,2,4,null,null,1], k = 3
// 输出：3

// 提示：

// 树中的节点数为 n 。
// 1 <= k <= n <= 104
// 0 <= Node.val <= 104

// 进阶：如果二叉搜索树经常被修改（插入/删除操作）并且你需要频繁地查找第 k 小的值，你将如何优化算法？
// 方法一：使用中序遍历
class Solution1
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

    int count = 0;
    int result = -1;

public:
    int kthSmallest(TreeNode *root, int k)
    {
        inorder(root, k);
        return result;
    }

private:
    void inorder(TreeNode *root, int k)
    {
        if (root == nullptr || count >= k)
            return;
        inorder(root->left, k);
        count++;
        if (count == k)
            result = root->val;
        inorder(root->right, k);
    }
};

// 方法二：使用手动栈实现中序遍历
class Solution2
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

public:
    int kthSmallest(TreeNode *root, int k)
    {
        stack<TreeNode *> stack;
        TreeNode *current = root;
        while (!stack.empty() || current != nullptr)
        {
            while (current != nullptr)
            {
                stack.push(current);
                current = current->left;
            }
            current = stack.top();
            stack.pop();
            k--;
            if (k == 0)
                return current->val;
            current = current->right;
        }
        return -1; // This line should never be reached
    }
};

// 方法三：子树返回其节点个数
class Solution2
{
private:
    struct TreeNode
    {
        int val;
        TreeNode *left;
        TreeNode *right;
        TreeNode() : val(0), left(nullptr), right(nullptr) {}
        TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
        TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    };

private:
    unordered_map<TreeNode *, int> node_num_map;

public:
    int kthSmallest(TreeNode *root, int k)
    {
        countNodeNum(root);
        return findKthSmallest(root, k);
    }

private:
    int findKthSmallest(TreeNode *root, int k)
    {
        TreeNode *current = root;
        while (current != nullptr)
        {
            int leftTreeNodeNum = getNodeNum(current->left);
            if (leftTreeNodeNum + 1 == k)
                return current->val;
            else if (leftTreeNodeNum + 1 < k)
            {
                k -= leftTreeNodeNum + 1;
                current = current->right;
            }
            else
                current = current->left;
        }
    }
    int countNodeNum(TreeNode *node)
    {
        if (node == nullptr)
            return 0;
        node_num_map[node] = 1 + countNodeNum(node->left) + countNodeNum(node->right);
        return node_num_map[node];
    }

    int getNodeNum(TreeNode *node)
    {
        if (node != nullptr && node_num_map.count(node))
            return node_num_map[node];
        return 0;
    }
};

// 方法四：AVL树

File: hot100/Solution0230.java
package hot100;

import java.util.ArrayDeque;
import java.util.Deque;
// 给定一个二叉搜索树的根节点 root ，和一个整数 k ，请你设计一个算法查找其中第 k 小的元素（k 从 1 开始计数）。

public class Solution0230 {
    private static class TreeNode {
        int val;
        int size; // 以当前节点为根的子树的节点总数
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
            this.size = 1; // 初始化size为1，表示当前节点本身
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.size = 1 + (left != null ? left.size : 0) + (right != null ? right.size : 0); // 计算size
            this.left = left;
            this.right = right;
        }
    }

    private int count = 0;
    private int result = 0;

    public int kthSmallest(TreeNode root, int k) {
        inorder(root, k);
        return result;
    }

    private void inorder(TreeNode node, int k) {
        if (node == null) {
            return;
        }
        inorder(node.left, k);
        count++;
        if (count == k) {
            result = node.val;
            return;
        }
        inorder(node.right, k);
    }

    private int kthSmallest2(TreeNode root, int k) {
        int count = 0;
        if (root == null) {
            return -1;
        }
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode current = root;
        while (current != null || !stack.isEmpty()) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }
            current = stack.pop();
            count++;
            if (count == k) {
                return current.val;
            }
            current = current.right;
        }
        return -1;
    }

    private int kthSmallest3(TreeNode root, int k) {
        int count = countNodes(root.left);
        if (k <= count) {
            return kthSmallest3(root.left, k);
        } else if (k > count + 1) {
            return kthSmallest3(root.right, k - count - 1);
        } else {
            return root.val;
        }
    }

    private int countNodes(TreeNode node) {
        if (node == null) {
            return 0;
        }
        return 1 + countNodes(node.left) + countNodes(node.right);
    }

    private int kthSmallest4(TreeNode root, int k) {
        if (root == null) {
            return -1;
        }
        int leftSize = root.left != null ? root.left.size : 0;
        if (k <= leftSize) {
            return kthSmallest4(root.left, k);
        } else if (k > leftSize + 1) {
            return kthSmallest4(root.right, k - leftSize - 1);
        } else {
            return root.val;
        }
    }

    public static void main(String[] args) {
        Solution0230 solution = new Solution0230();
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(1);
        root.right = new TreeNode(4);
        root.left.right = new TreeNode(2);
        int k = 1;
        System.out.println(solution.kthSmallest3(root, k)); // 输出: 1
    }
}
